### PR TITLE
Add AP_Perf Library (v2)

### PR DESCRIPTION
**Pull request by lucasdemarchi**
Rebased version of #6521

This still suffers from the bug fixed in #6704  - not sure which one should go in first.

And I still don't agree that we should disable the internal perf counters when there's lttng around. Lttng events should be generated in addition to the internal ones. In other words:  AP_Perf_Lttng is embedded in  AP_Perf_Linux and is called when lttng is around.

### CODE REVIEW

**Comment by lucasdemarchi** [comment]
I added some changes on top to do the adjustments I talked about. I'll do some tests and push tomorrow.

File: libraries/AP_Perf/AP_Perf.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+void AP_Perf::end(perf_counter_t pc)
+{
+    _backend->end(pc);
+}
+
+void AP_Perf::count(perf_counter_t pc)
+{
+    _backend->count(pc);
+}
```
Comment (khancyr) [inline libraries/AP_Perf/AP_Perf.cpp]: miss endline ^^

File: libraries/AP_Perf/AP_Perf_Linux.cpp [hunk 2]
```diff
@@ -104,20 +92,20 @@ Perf::Perf()
     _perf_counters.reserve(50);
 
 #ifdef DEBUG_PERF
-    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&Perf::_debug_counters, void));
+    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&AP_Perf_Lttng::_debug_counters, void));
```
Comment (khancyr) [inline libraries/AP_Perf/AP_Perf_Linux.cpp]: AP_Perf_Linux no ?

File: libraries/AP_Perf/AP_Perf_Linux.cpp [hunk 3]
```diff
@@ -92,7 +86,7 @@ AP_Perf_Linux::AP_Perf_Linux()
     _perf_counters.reserve(50);
 
 #ifdef DEBUG_PERF
-    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&AP_Perf_Lttng::_debug_counters, void));
+    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&AP_Perf_Linux::_debug_counters, void));
```
Comment (khancyr) [inline libraries/AP_Perf/AP_Perf_Linux.cpp]: ha corrected here!

**Comment by khancyr** [comment]
Miss the blank line at the end of plenty files.
I could be interesting to allow the same as linux for SITL. I think some people could be interested

File: libraries/AP_Perf/AP_Perf.cpp (hunk 1 above)
Comment (lucasdemarchi) [reply libraries/AP_Perf/AP_Perf.cpp]: Why a newline us needed here?

File: libraries/AP_Perf/AP_Perf_Linux.cpp (hunk 3 above)
Comment (lucasdemarchi) [reply libraries/AP_Perf/AP_Perf_Linux.cpp]: Yep, I took original commits as is and added changes on top, including this

**Comment by lucasdemarchi** [comment]
> I could be interesting to allow the same as linux for SITL. I think some people could be interested

Sure. I think it is a very simple change... Just changing the backend would probably be enough.

**Comment by lucasdemarchi** [comment]
Fixed the conflicts and added a commit on top to allow using it for SITL. @khancyr can you confirm if it works for you?

**Comment by lucasdemarchi** [comment]
Too many concflicts to solve, a new platform to support. It's better to have the **idea** to be redone from scratch rather than rebasing. Closing.

**Pull request closed without merging**